sled-agent/src/services.rs

```diff
@@ -4675,6 +4692,151 @@ impl ServiceManager {
     }
 }
 
+#[derive(Debug)]
+struct ReconciledNewZonesRequest {
+    zones_to_be_removed: HashSet<OmicronZoneConfig>,
+    zones_to_be_added: HashSet<OmicronZoneConfig>,
+}
+
+fn reconcile_running_zones_with_new_request<'a>(
+    existing_zones_with_runtime_zpool: impl Iterator<
+        Item = (&'a mut OmicronZoneConfig, &'a ZpoolOrRamdisk),
+    >,
+    new_request: OmicronZonesConfig,
+    log: &Logger,
+) -> Result<ReconciledNewZonesRequest, Error> {
+    let mut existing_zones_by_id: BTreeMap<_, _> =
+        existing_zones_with_runtime_zpool
+            .map(|(zone, zpool)| (zone.id, (zone, zpool)))
+            .collect();
+    let mut zones_to_be_added = HashSet::new();
+    let mut zones_to_be_removed = HashSet::new();
+    let mut zones_to_update = Vec::new();
+
+    for zone in new_request.zones.into_iter() {
+        let Some((existing_zone, runtime_zpool)) =
+            existing_zones_by_id.remove(&zone.id)
+        else {
+            // This zone isn't in the existing set; add it.
+            zones_to_be_added.insert(zone);
+            continue;
+        };
+
+        // We're already running this zone. If the config hasn't changed, we
+        // have nothing to do.
+        if zone == *existing_zone {
+            continue;
+        }
+
+        // Special case for fixing #7229. We have an incoming request for a zone
+        // that we're already running except the config has changed; normally,
+        // we'd shut the zone down and restart it. However, if we get a new
+        // request that is:
+        //
+        // 1. setting `filesystem_pool`, and
+        // 2. the config for this zone is otherwise identical, and
+        // 3. the new `filesystem_pool` matches the pool on which the zone is
+        //    installed
+        //
+        // then we don't want to shut the zone down and restart it, because the
+        // config hasn't actually changed in any meaningful way; this is just
+        // reconfigurator correcting #7229.
```

Comment:
More just out of curiosity than anything: Are there other situations where this is true? Why is #7229 special?

Reply:
Thanks for the context. What I was thinking about reading this comment (from the perspective of a future reader) was how I would know what other cases might be special besides 7229. I think the last sentence sort of gets at the more general "why"; maybe that could be moved up into a summary toward the top of the comment, but given that discussion is captured in the PR that's probably fine for now. Your call!